app/src/main/java/org/oppia/android/app/translation/AppLanguageWatcherMixin.kt

```diff
@@ -61,11 +65,34 @@ class AppLanguageWatcherMixin @Inject constructor(
       appLanguageLocaleHandler.initializeLocale(defaultDisplayLocale)
     }
 
-    // TODO(#52): Hook this up properly to profiles, and handle the non-profile activity cases.
-    val profileId = ProfileId.getDefaultInstance()
-    val appLanguageLocaleDataProvider = translationController.getAppLanguageLocale(profileId)
-    val liveData = appLanguageLocaleDataProvider.toLiveData()
-    liveData.observe(
+    val currentUserProfileId = profileManagementController.getCurrentProfileId()
+
+    var appLanguageLocaleDataProvider: DataProvider<OppiaLocale.DisplayLocale>? = null
+
+    when {
+      shouldUseSystemLanguage -> {
+        appLanguageLocaleDataProvider = translationController.getSystemLanguageLocale()
+      }
+      else -> {
+        when {
+          shouldUseSystemLanguage && currentUserProfileId == null -> {
+            appLanguageLocaleDataProvider = translationController.getSystemLanguageLocale()
+          }
+          !shouldUseSystemLanguage && currentUserProfileId != null -> {
+            appLanguageLocaleDataProvider =
+              translationController.getAppLanguageLocale(currentUserProfileId)
+          }
+          !shouldUseSystemLanguage && currentUserProfileId == null -> {
+            // Use default instance mostly in test scenarios where currentUserProfileId is always null.
+            appLanguageLocaleDataProvider =
+              translationController.getAppLanguageLocale(ProfileId.getDefaultInstance())
+          }
+        }
+      }
+    }
```

Comment:
```suggestion
    val appLanguageLocaleDataProvider = when {
      shouldUseSystemLanguage -> translationController.getSystemLanguageLocale()
      currentUserProfileId == null -> translationController.getSystemLanguageLocale()
      else -> translationController.getAppLanguageLocale(currentUserProfileId)
    }
```

This logic can be majorly simplified. And, upon simplifying, it shows that one of the cases isn't even possible.

Also, you can revert the nullable LiveData bit below since this ensures the provider is always present.

Reply:
Updated, however as expressed in test cases ProfileId is always null, will test this since it is what we want to happen in normal situation of the system in Action.